src/main/java/com/google/devtools/build/lib/analysis/starlark/UnresolvedSymlinkAction.java

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.devtools.build.lib.analysis.starlark;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.devtools.build.lib.actions.AbstractAction;
+import com.google.devtools.build.lib.actions.ActionExecutionContext;
+import com.google.devtools.build.lib.actions.ActionExecutionException;
+import com.google.devtools.build.lib.actions.ActionKeyContext;
+import com.google.devtools.build.lib.actions.ActionOwner;
+import com.google.devtools.build.lib.actions.ActionResult;
+import com.google.devtools.build.lib.actions.Artifact;
+import com.google.devtools.build.lib.actions.Artifact.ArtifactExpander;
+import com.google.devtools.build.lib.analysis.actions.SymlinkAction;
+import com.google.devtools.build.lib.collect.nestedset.NestedSetBuilder;
+import com.google.devtools.build.lib.collect.nestedset.Order;
+import com.google.devtools.build.lib.server.FailureDetails;
+import com.google.devtools.build.lib.server.FailureDetails.FailureDetail;
+import com.google.devtools.build.lib.server.FailureDetails.SymlinkAction.Code;
+import com.google.devtools.build.lib.util.DetailedExitCode;
+import com.google.devtools.build.lib.util.Fingerprint;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.lib.vfs.PathFragment;
+import java.io.IOException;
+import javax.annotation.Nullable;
+
+/**
+ * Action to create a possibly unresolved symbolic link to a raw path.
+ *
+ * <p>To create a symlink to a known-to-exist target with alias semantics similar to a true copy of
+ * the input, use {@link SymlinkAction} instead.
+ */
+public final class UnresolvedSymlinkAction extends AbstractAction {
+  private static final String GUID = "0f302651-602c-404b-881c-58913193cfe7";
+
+  private final String target;
```

Comment:
Why is this not a `PathFragment`? (it looks like it's a little bit more economical with GC if the action is invoked more than once during the lifetime of a server)

I mean, it should probably be a String because `PathFragment` is normalized and that's surprising, but it'd be a bit more involved to change that since we can currently only create symlinks from `PathFragment`s in the VFS layer, but for now, why not turn it into a `PathFragment` as early as possible?

Reply:
Changed and moved up the comment. I intend to resolve this in a follow-up effort, but agree that using `PathFragment` is better for now. I would prefer to keep the `String` parameter of the constructor though as that's the API we ultimately want to realize.

Reply:
As you desire. Given the limitation of VFS it doesn't matter much anyways.

Reply:
I actually had that limitation removed on a branch and it wasn't so painful. I will revisit this, but would first want to understand the level of Windows support we can achieve.